src/UniswapV4MulticurveRehypeInitializerHook.sol

```diff
@@ -0,0 +1,687 @@
+            balance0 = isToken0 ? balance0 : balance0 - assetBuybackAmountIn;
+            balance1 = isToken0 ? balance1 - assetBuybackAmountIn : balance1;
+        }
+
+        if (numeraireBuybackAmount > 0) {
+            Currency outputCurrency = isToken0 ? key.currency1 : key.currency0;
+            // must check poolManager balance or else this will revert on failed transfer due to missing tokens
+            if (IERC20(Currency.unwrap(outputCurrency)).balanceOf(address(poolManager)) > numeraireBuybackAmount) {
+                (, uint256 numeraireBuybackAmountOut, uint256 numeraireBuybackAmountIn) =
+                    _executeSwap(key, isToken0, numeraireBuybackAmount);
+                isToken0
+                    ? key.currency1.transfer(getPoolInfo[poolId].buybackDst, numeraireBuybackAmountOut)
+                    : key.currency0.transfer(getPoolInfo[poolId].buybackDst, numeraireBuybackAmountOut);
+                balance0 = isToken0 ? balance0 - numeraireBuybackAmountIn : balance0;
+                balance1 = isToken0 ? balance1 : balance1 - numeraireBuybackAmountIn;
+            }
+        }
+        (uint160 sqrtPriceX96,,,) = poolManager.getSlot0(poolId);
+
+        Position storage position = getPosition[poolId];
+        _initializePositionIfNeeded(position, key, poolId);
+
+        bool shouldSwap;
+        bool zeroForOne;
+        uint256 swapAmountIn;
+        uint256 swapAmountOut;
+        uint160 postSwapSqrtPrice;
+
+        (shouldSwap, zeroForOne, swapAmountIn, swapAmountOut, postSwapSqrtPrice) =
+            _rebalanceFees(key, lpAmount0, lpAmount1, sqrtPriceX96);
+
+        if (shouldSwap && swapAmountIn > 0) {
+            Currency outputCurrency = zeroForOne ? key.currency1 : key.currency0;
+            if (IERC20(Currency.unwrap(outputCurrency)).balanceOf(address(poolManager)) > swapAmountOut) {
+                (postSwapSqrtPrice, swapAmountOut, swapAmountIn) = _executeSwap(key, zeroForOne, swapAmountIn);
+                lpAmount0 = zeroForOne ? lpAmount0 - swapAmountIn : lpAmount0 + swapAmountOut;
+                lpAmount1 = zeroForOne ? lpAmount1 + swapAmountOut : lpAmount1 - swapAmountIn;
+                (uint256 amount0Added, uint256 amount1Added) =
+                    _addFullRangeLiquidity(key, position, lpAmount0, lpAmount1, postSwapSqrtPrice);
+                uint256 remainder0 = lpAmount0 - amount0Added;
+                uint256 remainder1 = lpAmount1 - amount1Added;
+                balance0 = zeroForOne ? balance0 - (amount0Added + swapAmountIn) + remainder0 : balance0 + remainder0;
+                balance1 = zeroForOne ? balance1 + remainder1 : balance1 - (amount1Added + swapAmountIn) + remainder1;
+            }
+        }
+
+        getHookFees[poolId].beneficiaryFees0 += uint128(balance0);
+        getHookFees[poolId].beneficiaryFees1 += uint128(balance1);
+
+        getHookFees[poolId].fees0 = 0;
+        getHookFees[poolId].fees1 = 0;
```

Comment:
Something to consider is that using this pattern of doing buybacks before rehypothecating and the collecting the remainder as the beneficiary fees means that depending on the params used we might rehypothecate very rarely. To elaborate, any time the fees exceed EPSILON, we do buybacks before rehypothecating only if we still have enough excess fees, then if we don't rehypothecate, we just collect the remainder as beneficiary fees. So we only ever actually rehypothecate if we exceed the EPSILON amount in one swap by enough that we can cover buybacks and still have enough remaining to also swap/rehype